packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts

```diff
@@ -83,6 +83,11 @@ const result2 = foo() == null;
     necessaryConditionTest('null | object'),
     necessaryConditionTest('undefined | true'),
     necessaryConditionTest('void | true'),
+    // "branded" type
+    necessaryConditionTest('string & {}'),
+    necessaryConditionTest('string & { __brand: string }'),
+    necessaryConditionTest('number & {}'),
+    necessaryConditionTest('boolean & {}'),
```

Comment:
[Testing] Normally we'll want both `invalid` and `valid` cases added. This just adds `valid`. Could you add some more coverage?

Additionally, some maybe missing areas:

* `bigint`
* ...I don't think symbol brands are a thing, so maybe it's fine to skip them?
* `{} & { __brand: string }`: should this still cause a report?
* `{ a: number } & { b: string }`: should this still cause a report?
* `string & string`

Reply:
Thanks for the feedback! 😃

I have added more tests, especially for more complex scenarios that looked like potential edge cases to me.

While adding tests for `invalid` cases, I noticed that `isPossiblyTruthy` also needed to be fixed: If we have a scenario like `"" & { __brand: string }`, the rule would not report any error. Thus, we need to look at the intersection members and make sure that all of them are not falsy.  
(Interesting side note: The rule as currently found on `main` would report `alwaysTruthy`.)

---

Two points regarding your comments:

> I don't think symbol brands are a thing, so maybe it's fine to skip them?

I think so too 👍 

> `{} & { __brand: string }` and `{ a: number } & { b: string }`

I believe that both should behave just like if you wrote `{ __brand: string }`, since the first is equal to `{ __brand: string }`, and the second is equal to `{ a: number, b: string }`.  
Thus, `{ a: number } & { b: string } & string` should not cause a report. But `{ a: number } & { b: string } & "foo"` should cause a report.


---

Btw: I tried to find a scenario where we need to recursively unwrap unions/intersections. Something like `foo & (bar | (baz & frub))`. But I simply couldn't come up with one where the type checker actually reports it in this particular shape. Or, in other words: I couldn't find a scenario where the current implementation would fail. Thus, I decided not to add any recursive logic for now.